cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -0,0 +1,321 @@
+                    let fileDir = takeDirectory file
+                        perm' = case perm of
+                            Exec -> Tar.executableFilePermissions
+                            NoExec -> Tar.ordinaryFilePermissions
+                    needsEntry <- gets (Set.notMember fileDir)
+
+                    when needsEntry $ do
+                        modify (Set.insert fileDir)
+                        case Tar.toTarPath True fileDir of
+                            Left err -> liftIO $ die' verbosity ("Error packing sdist: " ++ err)
+                            Right path -> tell [Tar.directoryEntry path]
+                        
+                    contents <- liftIO $ BSL.readFile file
+                    case Tar.toTarPath False file of
+                        Left err -> liftIO $ die' verbosity ("Error packing sdist: " ++ err)
+                        Right path -> tell [(Tar.fileEntry path contents) { Tar.entryPermissions = perm' }]
+            
+            entries <- execWriterT (evalStateT entriesM mempty)
+            let -- Pretend our GZip file is made on Unix.
+                normalize bs = BSL.concat [first, "\x03", rest']
+                    where
+                        (first, rest) = BSL.splitAt 9 bs
+                        rest' = BSL.tail rest
+            write . normalize . GZip.compress . Tar.write $ entries
+            when (outputFile /= "-") $
+                notice verbosity $ "Wrote tarball sdist to " ++ outputFile ++ "\n"
+        Archive ZipFormat -> do
+            entries <- forM files $ \(perm, file) -> do
+                let perm' = case perm of
+                        -- -rwxr-xr-x
+                        Exec   -> 0o010755 `shiftL` 16
+                        -- -rw-r--r--
+                        NoExec -> 0o010644 `shiftL` 16
+                contents <- BSL.readFile file
+                return $ (Zip.toEntry file 0 contents) { Zip.eExternalFileAttributes = perm' }
+            let archive = foldr Zip.addEntryToArchive Zip.emptyArchive entries
+            write (Zip.fromArchive archive)
+            when (outputFile /= "-") $
+                notice verbosity $ "Wrote zip sdist to " ++ outputFile ++ "\n"
+
+--
+
+reifyTargetSelectors :: [PackageSpecifier UnresolvedSourcePackage] -> [TargetSelector] -> Either [TargetProblem] [UnresolvedSourcePackage]
+reifyTargetSelectors pkgs sels = 
+    case partitionEithers (foldMap go sels) of
+        ([], sels') -> Right sels'
+        (errs, _)   -> Left errs
+    where
+        flatten (SpecificSourcePackage pkg@SourcePackage{}) = pkg
+        flatten _ = error "The impossible happened: how do we not know about a local package?"
+        pkgs' = fmap flatten pkgs
+
+        getPkg pid = case find ((== pid) . packageId) pkgs' of
+            Just pkg -> Right pkg
+            Nothing -> error "The impossible happened: we have a reference to a local package that isn't in localPackages."
```

Comment:
Perhaps add a `TargetProblemImpossible` and use it instead of `error` here?

Reply:
Didn't do it since it's harder to get rid of the one in flatten and I don't feel like it's worth the work when there's another five lines above.